store/bolt/user_store.go

```diff
@@ -0,0 +1,124 @@
+package bolt
+
+import (
+	"errors"
+
+	"github.com/boltdb/bolt"
+	"github.com/gopheracademy/congo"
+	"github.com/gopheracademy/congo/store/bolt/internal"
+)
+
+var (
+	// ErrUserNotFound is returned when mutating a user that doesn't exist.
+	ErrUserNotFound = errors.New("user not found")
+
+	// ErrEmailRequired is returned when creating a user without an email.
+	ErrEmailRequired = errors.New("email address required")
+)
+
+// UserStore represents the data store for user data.
+type UserStore struct {
+	db *bolt.DB
+}
+
+// initialize creates the top-level buckets for the store.
+func (s *UserStore) init() error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		tx.CreateBucketIfNotExists([]byte("users"))
+		tx.CreateBucketIfNotExists([]byte("users.email"))
+		return nil
+	})
+}
+
+// User returns a user by ID.
+func (s *UserStore) User(id uint64) (u *congo.User, err error) {
+	err = s.db.View(func(tx *bolt.Tx) error {
+		if v := tx.Bucket([]byte("users")).Get(u64tob(id)); v != nil {
+			u, err = internal.UnmarshalUser(v)
+			return err
+		}
+		return nil
+	})
+	return
+}
```

Comment:
Because Bolt is typically accessed using the `View()` and `Update()` functions which require closures, there's not an easy way to return data from inside the closure to the outside function. This works although some people may think it's hacky at first. The `Get()` on L36 retrieves the protobuf encoded bytes for the given user ID. If there are bytes there then it unmarshals those bytes on L37 into `u` which is the return variable declared on L34. We then return any error that occurred which gets passed through the `View()` and reassigned to the `err` return variable on L36. Finally, we `return` on L42 with whatever values are in `u` and `err`.

Although this sounds confusing it cuts down on a bunch of extra lines and noise you'd have to do otherwise if you didn't use named return variables.